### PR TITLE
Add investment projects activity / SPI report dataset endpoint

### PR DESCRIPTION
**Pull request by elcct**
### Description of change

The `GET /v4/dataset/investment-projects-activity-dataset` has been added. The endpoint returns SPI report records for corresponding investment projects.

It updates the SPI report so that formatting of propositions can be changed, as the response of dataset is in JSON, so there is no need for `;` separated objects.

There will be a second PR that removes SPI report from Django admin. That should resolve duplicated `ist_adviser` and `propositions` fixtures in dataset app.

For a time being we will have to run both versions of the report (the endpoint and celery task). Once we remove the task and no longer need to support both looks, the implementation should become much cleaner.

### Checklist

* [x] Has a new newsfragment been created? Check [changelog/README.md](https://github.com/uktrade/data-hub-api/blob/master/changelog/README.md) for instructions
* [ ] Do any added or updated endpoints appear in the API documentation? See [docs/Maintaining the API documentation.md](https://github.com/uktrade/data-hub-api/blob/develop/docs/Maintaining&#32;the&#32;API&#32;documentation.md) for more details
* [ ] Have any relevant search models been updated?
* [ ] Have any relevant fixtures (`fixtures/test_data.yaml`) been updated?
* [ ] Have any relevant select-/prefetch-related field lists in the views and search apps been updated?
* [ ] Has the admin site been updated (for new models, fields etc.)?
* [ ] Has the README been updated (if needed)?

### CODE REVIEW

**Comment by reupen** [comment]
@elcct I gave this a try locally.

First thing is that, for a page of 100 projects, it is executing over 300 queries:

image

My database is very bare so it is hard to know how that will behave in production. Can it be optimised?

Second, the response looks like this:

```json
{
  "next": "http://localhost:8000/v4/dataset/investment-projects-activity-dataset?cursor=cD1mN2FiZjA3NS01MDcxLTQ2YzYtYTMyZS03ZTM1MzhmZWNiY2I%3D&format=json",
  "previous": null,
  "results": [
    {
      "investment_project_id": "00195fc6-d1c1-4478-87f0-b57e67e65ec1"
    },
    {
      "investment_project_id": "0283955a-c9bc-45e5-b6c6-ab4a42122fbc"
    },
    {
      "investment_project_id": "07f41b5f-3adc-4f9f-9ccc-3c1c30596bc4"
    },
    {
      "investment_project_id": "0c4c31dd-ebec-42ff-889a-c80963ee6168"
    }
  ]
}
```

Are the result properties dynamic?

**Comment by elcct** [comment]
@reupen It does make at least one query per investment project, that's why the number is quite high. This, however, works the same as the original report. This is also not meant to be called frequently - ideally once a day. I am looking into how I could improve the performance. 

The fields are created if corresponding value exists.

**Comment by reupen** [comment]
> This, however, works the same as the original report.

However, the calling context is different as this is running in a web process and not Celery. I'm also not sure if there would be even more queries with more realistic data.

> The fields are created if corresponding value exists.

This doesn't seem like standard behaviour – could you elaborate on the reason for it? Does it fit in with how Data Workspace works?

**Comment by elcct** [comment]
Fair point re calling context. I am going to try to move it into a single query. 

Re dynamic fields, I see that this is wrong - in case of CSV output, that would mean just an empty cell, but it doesn't make sense for JSON object. I am going to change that. Thank you!

**Comment by elcct** [comment]
I have improved the report performance. For my 100 investment projects, the number of queries went down from 317 to 17. I am looking at ways to improve it further.
The endpoint is now returning all fields as well.

**Comment by elcct** [comment]
I have updated the query and now it does just one.

File: datahub/dataset/investment_project/pagination.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+from datahub.investment.project.proposition.models import PropositionStatus
+from datahub.investment.project.report.spi import SPIReport
+
+
+class InvestmentProjectActivityDatasetViewCursorPagination(DatasetCursorPagination):
+    """Cursor Pagination for SPI Report."""
+
+    ordering = ('id', )
+
+    required_fields_label_mapping = {
```
Comment (currycoder) [inline datahub/dataset/investment_project/pagination.py]: Could there perhaps be some comments explaining the purpose of these two class attributes?

File: datahub/dataset/investment_project/pagination.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+from dateutil.parser import parse as dateutil_parse
+from rest_framework.response import Response
+
+from datahub.dataset.core.pagination import DatasetCursorPagination
+from datahub.investment.project.proposition.models import PropositionStatus
+from datahub.investment.project.report.spi import SPIReport
+
+
+class InvestmentProjectActivityDatasetViewCursorPagination(DatasetCursorPagination):
```
Comment (currycoder) [inline datahub/dataset/investment_project/pagination.py]: It's a little confusing to me that this pagination class has so much formatting logic within it.  I think it could be cleaner with neater separation of concerns if there was an additional `SPIReportFormatter` class of some kind which `get_paginated_response()` uses to build formatted results from data.

That might also allow for some extended comments/class documentation which explain how the translation between the SPIReport and the chosen formatting is achieved.  The formatter logic could also be tested in isolation.

File: datahub/investment/project/report/spi.py [hunk 3]
```diff
@@ -171,16 +166,15 @@ def _format_propositions(self, propositions):
         deadline;ongoing;;adviser.name;deadline;completed;modified_on;adviser.name...
         """
         formatted = []
-
         for proposition in propositions:
-            formatted.append(proposition.deadline.isoformat())
-            formatted.append(proposition.status)
-            if proposition.status == PropositionStatus.ongoing:
+            formatted.append(dateutil_parse(proposition['deadline']).strftime('%Y-%m-%d'))
```
Comment (currycoder) [inline datahub/investment/project/report/spi.py]: Could you help me understand when this default proposition formatter might be used as opposed to `self.proposition_formatter`?  e.g. what business cases are each used for..?

Comment (elcct) [reply datahub/investment/project/report/spi.py]: The default formatter is being used in the CSV report that is generated by the Celery task. I understand it is weird, but that was the requirement for the propositions to look in the spreadsheet cell. Once the new endpoint is in use, we will be able to remove that.

File: datahub/dataset/investment_project/pagination.py (hunk 2 above)
Comment (elcct) [reply datahub/dataset/investment_project/pagination.py]: I agree that this is not ideal and thank you for suggestions. I'll have a look how to improve it.

File: datahub/investment/project/report/spi.py (hunk 3 above)
Comment (currycoder) [reply datahub/investment/project/report/spi.py]: Ok I understand - thanks.

I think this leads me to my next thought. It feels a little bit like some of the complexity here is because of this situation: We have this existing unit of code whose contract is to generate CSV formatted SPI reports by interrogating the DB and formatting results.  Some of the complexity is from transforming this CSV output in to something that can work on an API endpoint.  Could we instead leave the legacy CSV implementation as it is and develop some new units which 1) grab the data 2) format it, and 3) expose it on an endpoint for DW

I realise that what I'm saying represents a big refactor and the current implementation works etc.  I'm just wondering whether we will end up in a situation where we are stuck with this CSV formatting code that won't ultimately make sense when we are completely using an API?

Comment (elcct) [reply datahub/investment/project/report/spi.py]: The end goal is to remove the CSV generating code and associated formatting, so it could greatly reduce the complexity. The reason I re-used the existing code is because I wanted to reduce possibility of having its output different from the (to be) legacy report, as it is important that the output is exactly the same (minus a couple of fields that are no longer needed and the formatting of propositions). So I'd rather not create a second implementation. I'd prefer to follow this PR with another that will remove any (to be) legacy code instead.

**Comment by elcct** [comment]
Currently this PR is blocked until https://github.com/uktrade/data-hub-api/pull/2486 is merged, because with the current version of Django that is being used, the `ArrayAgg(filter=...)` doesn't work.

File: datahub/investment/project/report/spi.py (hunk 3 above)
Comment (currycoder) [reply datahub/investment/project/report/spi.py]: Ok awesome - I understand your rationale here and glad to hear that the CSV generating code will eventually be ripped out.

File: datahub/core/query_utils.py [hunk 4]
```diff
@@ -124,7 +125,12 @@ def get_array_agg_subquery(
     """
     return get_aggregate_subquery(
         model,
-        ArrayAgg(expression_to_aggregate, distinct=distinct, ordering=ordering),
+        ArrayAgg(
+            expression_to_aggregate,
+            distinct=distinct,
+            ordering=ordering,
+            filter=Q(**filter) if filter else None,
```
Comment (reupen) [inline datahub/core/query_utils.py]: I think it would be better to just pass `filter` through (so the caller can pass a Q object directly).

File: datahub/dataset/investment_project/pagination.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+    def _proposition_formatter(self, propositions):
+        """Returns a list of propositions with selected fields."""
+        return [{
+            'deadline': dateutil_parse(proposition['deadline']).strftime('%Y-%m-%d'),
+            'status': proposition['status'],
+            'modified_on': dateutil_parse(proposition['modified_on']).isoformat()
+            if proposition['status'] != PropositionStatus.ongoing else '',
+            'adviser_id': proposition['adviser_id'],
+        } for proposition in propositions]
```
Comment (reupen) [inline datahub/dataset/investment_project/pagination.py]: Would be good to try and improve the formatting here e.g.:

```python
        return [
            {
                'deadline': dateutil_parse(proposition['deadline']).strftime('%Y-%m-%d'),
                'status': proposition['status'],
                'modified_on':
                    dateutil_parse(proposition['modified_on']).isoformat()
                    if proposition['status'] != PropositionStatus.ongoing else '',
                'adviser_id': proposition['adviser_id'],
            }
            for proposition in propositions
        ]
```

File: datahub/dataset/investment_project/pagination.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+    def get_paginated_response(self, data):
+        """Get paginated response."""
+        data = self._get_data(data)
+        results = self._map_results(data)
+        return Response(OrderedDict([
+            ('next', self.get_next_link()),
+            ('previous', self.get_previous_link()),
+            ('results', results),
+        ]))
```
Comment (reupen) [inline datahub/dataset/investment_project/pagination.py]: Should be able to use a dict literal here (instead of `OrderedDict`) since we don't support old versions of Python.

File: datahub/core/query_utils.py (hunk 4 above)
Comment (elcct) [reply datahub/core/query_utils.py]: Good point, that should be responsibility of the caller. Updated. Thank you!

File: datahub/dataset/investment_project/pagination.py (hunk 6 above)
Comment (elcct) [reply datahub/dataset/investment_project/pagination.py]: Good point! Updated. Thank you!

File: datahub/dataset/investment_project/pagination.py (hunk 5 above)
Comment (elcct) [reply datahub/dataset/investment_project/pagination.py]: Yes, that will be an improvement! Updated, thank you!

File: datahub/dataset/investment_project/pagination.py (hunk 2 above)
Comment (elcct) [reply datahub/dataset/investment_project/pagination.py]: I moved the formatting code into a separate `SPIReportFormatter` class. It looks much cleaner now. I'll have a look into adding more comments now.

**Comment by elcct** [comment]
Since Django 3.0.3 update has now been merged, I have rebased this PR so that aggregate queries can work correctly.

File: datahub/core/query_utils.py [hunk 7]
```diff
@@ -158,7 +173,6 @@ def get_aggregate_subquery(model, expression, join_field_name='pk'):
     ).values(
         '_annotated_value',
     )
-
```
Comment (reupen) [inline datahub/core/query_utils.py]: Is this change intended?

File: datahub/dataset/investment_project/pagination.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+    def get_paginated_response(self, data):
+        """Get paginated response."""
+        spi_report = SPIReportFormatter()
+        results = spi_report.format(data)
+        return Response({
+            'next': self.get_next_link(),
+            'previous': self.get_previous_link(),
+            'results': results,
+        })
```
Comment (reupen) [inline datahub/dataset/investment_project/pagination.py]: Minor point, but we should break before and after `{` and `}` if we're following the code style.

File: datahub/dataset/investment_project/spi.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    required_fields_value_mapping = {
+        'Enquiry processed by': lambda adviser: str(adviser),
+        'Project manager assigned by': lambda adviser: str(adviser.id),
+    }
+
+    def __init__(self):
+        """Initialise SPI Report Formatter."""
+        self._SPIReport = SPIReport(proposition_formatter=proposition_formatter)
```
Comment (reupen) [inline datahub/dataset/investment_project/spi.py]: Maybe it would be simpler to use `proposition_formatter` in this class instead of passing it through to `SPIReport`? (It seems to be adding extra complexity that we might not need.)

File: datahub/investment/project/report/spi.py [hunk 10]
```diff
@@ -259,7 +261,6 @@ def get_spi_data_for_investment_project(self, investment_project):
         data.update(self.get_spi2(investment_project, spi_interaction_dates))
         data.update(self.get_spi3(investment_project))
         data.update(self.get_spi5(investment_project, spi_interaction_dates))
-
```
Comment (reupen) [inline datahub/investment/project/report/spi.py]: Was this change intended?

File: datahub/investment/project/report/spi.py [hunk 11]
```diff
@@ -226,10 +225,13 @@ def get_spi3(self, investment_project):
...
         data = {}
 
-        propositions = self._get_propositions(investment_project, ist_only=True)
-        if propositions:
-            data[self.SPI3] = self._format_propositions(propositions)
+        if not investment_project.spi_propositions:
+            return data
+
+        formatter = self.proposition_formatter \
+            if self.proposition_formatter else self._format_propositions
```
Comment (reupen) [inline datahub/investment/project/report/spi.py]: Minor point – if this remains, would be good to wrap this in `(` and `)` instead of using `\` (to be in line with PEP 8).

File: datahub/dataset/investment_project/pagination.py [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+from rest_framework.response import Response
+
+from datahub.dataset.core.pagination import DatasetCursorPagination
+from datahub.dataset.investment_project.spi import SPIReportFormatter
+
+
+class InvestmentProjectActivityDatasetViewCursorPagination(DatasetCursorPagination):
+    """Cursor Pagination for SPI Report."""
+
+    ordering = ('id', )
```
Comment (reupen) [inline datahub/dataset/investment_project/pagination.py]: The cursor pagination is really meant to be ordered by e.g. `('created_on', 'pk')`. Is there a reason we didn't stick to that here?

File: datahub/dataset/investment_project/spi.py (hunk 9 above)
Comment (elcct) [reply datahub/dataset/investment_project/spi.py]: I think there is no access to raw propositions in the data returned by `SPIReport` , so I would have to parse `;` separated ones back into JSON if I was going to do it in this class. The goal is to remove the old formatting once people have switched to data returned by this endpoint, so that complexity will be removed then.

File: datahub/dataset/investment_project/pagination.py (hunk 12 above)
Comment (elcct) [reply datahub/dataset/investment_project/pagination.py]: Good point! There is no good reason - I have updated it to order by `created_on` and `pk`.

File: datahub/investment/project/report/spi.py (hunk 11 above)
Comment (elcct) [reply datahub/investment/project/report/spi.py]: Ah! Updated it now. Thank you!

File: datahub/investment/project/report/spi.py (hunk 10 above)
Comment (elcct) [reply datahub/investment/project/report/spi.py]: No. I have reversed it now. Thank you!

File: datahub/dataset/investment_project/pagination.py (hunk 8 above)
Comment (elcct) [reply datahub/dataset/investment_project/pagination.py]: I am too used to this formatting :( Updated it now. Thank you!

File: datahub/core/query_utils.py (hunk 7 above)
Comment (elcct) [reply datahub/core/query_utils.py]: No! I was too happy with deleting :-) Updated. Thank you.

File: datahub/dataset/investment_project/spi.py (hunk 9 above)
Comment (reupen) [reply datahub/dataset/investment_project/spi.py]: :+1:
